astroquery/nasa_exoplanet_archive/core.py

```diff
@@ -59,10 +73,39 @@
     "pi_rel": None,
     "ppm": cds.ppm,
     "seconds": u.s,
+    "Solar mass": u.M_sun,
     "solarradius": u.R_sun,
+    "Solar Radius": u.R_sun,
+    "log10(cm/s**2)": u.dex(u.cm / u.s ** 2),
+    "dex": u.dex(None),
+    "sexagesimal": None
 }
+
+# Converter for converting raw string values from the table into numeric data types. Dictionary items will be converters (converter function, converter type tuples) for consecutive columns.
 CONVERTERS = dict(koi_quarters=[ascii.convert_numpy(str)])
-OBJECT_TABLES = {"exoplanets": "pl_", "compositepars": "fpl_", "exomultpars": "mpl_"}
+
+# 'ps' and 'pscomppars' are the main tables of detected exoplanets. Calls to the old tables ('exoplanets', 'compositepars', 'exomultpars') will return errors and urge the user to call the 'ps' or 'pscomppars' tables
+OBJECT_TABLES = {"ps": "pl_", "pscomppars": "pl_", "exoplanets": "pl_", "compositepars": "fpl_", "exomultpars": "mpl_"}
+MAP_TABLEWARNINGS = {"exoplanets": "Planetary Systems (PS)", "compositepars": "Planetary System Composite Parameters table (PSCompPars)", "exomultpars": "Planetary Systems (PS)"}
+
+
+def get_access_url(service='tap'):
+    if service == 'tap':
+        url = conf.url_tap
+    elif service == 'api':
+        url = conf.url_api
+    return url
+
+
+def get_tap_tables(url):
+    """Tables accessed by API are gradually migrating to TAP service. Generate current list of tables in TAP."""
+    tap = pyvo.dal.tap.TAPService(baseurl=url)
+    response = tap.search(query="select * from TAP_SCHEMA.tables", language="ADQL")
+    tables = [table for table in response["table_name"].data if "TAP_SCHEMA." not in table]
+    return tables
+
+
+TAP_TABLES = get_tap_tables(conf.url_tap)  # TAP_TABLES = ['ps', 'pscomppars']
```

Comment:
This line is the problem: you can't (by astroquery rules) perform a remote query at import time.

Reply:
Thanks @keflavich, that makes perfect sense.

Reply:
@keflavich - I think we are pretty close now. TAP_TABLES is now an attribute of the class, and get_tap_tables is also a method of the class. Everything seems to work in terms of module functionality, only issue is the testing which now insists on trying to use the TAP service for all the tests, and failing with `AssertionError: assert 'https://exop...edu/TAP//sync' == 'https://exop.../nph-nstedAPI' `. It isn't clear to me why fetching the TAP table names in core.py introduces this change in `test_regularize_object_name`, `test_backwards_compat`, and `test_api_tables` . Any idea?

Reply:
Yeah, it looks like those tests have monkeypatched the `Session.request`, but they are calling functions that try to use the TAP URL and therefore call the `TAP_TABLES` attribute - so they're trying to send a request.   You probably need to monkeypatch the `TAP_TABLES` attribute in those tests.

Reply:
Ah, I see what you mean. Alright, let me see if I can get that fixed.

Reply:
My naive
```
    # Patch TAP_TABLES
    mp.setattr(__name__+".NasaExoplanetArchive.TAP_TABLES", property(lambda: ['notap']))
```
in `patch_get` didn't seem to work, and then I went down a very deep rabbit hole in which I'm still stuck. Is monkeypatching supposed to be this hard? I just want to set the TAP_TABLES property to an empty list in these tests of API tables. It seems like `mp.setattr` still runs the TAP query. Part of error message:

```
...
../../../../_pytest/monkeypatch.py:100: in derive_importpath
    annotated_getattr(target, attr, ann=module)
../../../../_pytest/monkeypatch.py:84: in annotated_getattr
    obj = getattr(obj, name)
../core.py:134: in TAP_TABLES
    self._tap_tables = get_tap_tables()
../core.py:103: in get_tap_tables
    response = tap.search(query="select * from TAP_SCHEMA.tables", language="ADQL")
...
```

Reply:
the problem is, I think, with that `__name__` thing - why are you doing that?

You can do something much simpler here - technically you don't even need to monkeypatch for this case:
```python
def test_regularize_object_name(patch_get):
    NasaExoplanetArchive._tap_tables = ['list']
```

but if you were going to rename an attribute, it would be more like:
```python
mp.setattr(NasaExoplanetArchive, '_tap_tables', ['list'])
```

Reply:
I'm wrong about the monkeypatching approach... not entirely sure what you would need to do for that, but it's something different here.  Possibly would need `lambda x: ['notap']` because it expects to receive `self` (it's a method on a class instance), but that might require going one level deeper... in any case, you don't need this here.

Reply:
The `__name__` thing was because I was getting a Module Not Found error, I assume due to the relative import of `NasaExoplanetArchive` from `..core`. I'll try your suggestion. Thanks!

Reply:
in this case, `NasaExoplanetArchive` is actually an _instance_ of a class everywhere you're using it, so you don't need to monkeypatch it.  Monkeypatching is for changing things that are harder to access.... for example, in this case, I think you actually wanted to patch `get_tap_tables`, e.g.:
```python
@patch('astroquery.nasa_exoplanet_archive.core.get_tap_tables',
       lambda x: ['notap'])
```

that seems to also work (and means you can test the `TAP_TABLES` attribute)

Reply:
Thanks @keflavich. The checks are finally all green (except Codecov not quite on target, but hopefully you can let that slide). Now I just have to fix that crazy first commit message, and we are set.

Reply:
Don't worry about codecov, its status is mostly used informally.